mpc-core/src/msgs/garble.rs

```diff
@@ -1,10 +1,330 @@
-use crate::garble::{BinaryLabel, EncryptedGate};
-use mpc_circuits::CircuitId;
+use std::{collections::HashSet, sync::Arc};
+
+use crate::{garble, Block};
+use mpc_circuits::{Circuit, CircuitId};
+
+#[derive(Debug, Clone)]
+pub enum GarbleMessage {
+    GarbledCircuit(GarbledCircuit),
+}
+
+#[derive(Debug, Clone)]
+pub struct InputLabels {
+    pub id: usize,
+    pub labels: Vec<Block>,
+}
+
+impl From<garble::InputLabels<garble::WireLabel>> for InputLabels {
+    fn from(labels: garble::InputLabels<garble::WireLabel>) -> Self {
+        Self {
+            id: labels.id(),
+            labels: labels
+                .as_ref()
+                .into_iter()
+                .map(|label| *label.as_ref())
+                .collect::<Vec<Block>>(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct OutputEncoding {
+    pub id: usize,
+    pub encoding: Vec<bool>,
+}
+
+impl From<garble::label::OutputLabelsEncoding> for OutputEncoding {
+    fn from(encoding: garble::label::OutputLabelsEncoding) -> Self {
+        Self {
+            id: encoding.output.id,
+            encoding: encoding
+                .as_ref()
+                .iter()
+                .copied()
+                .map(|enc| *enc)
+                .collect::<Vec<bool>>(),
+        }
+    }
+}
 
 #[derive(Debug, Clone)]
 pub struct GarbledCircuit {
     pub id: CircuitId,
-    pub input_labels: Vec<BinaryLabel>,
-    pub encrypted_gates: Vec<EncryptedGate>,
-    pub decoding: Option<Vec<bool>>,
+    pub input_labels: Vec<InputLabels>,
+    pub encrypted_gates: Vec<Block>,
+    pub encoding: Option<Vec<OutputEncoding>>,
+}
+
+impl From<garble::GarbledCircuit<garble::Partial>> for GarbledCircuit {
+    fn from(gc: garble::GarbledCircuit<garble::Partial>) -> Self {
+        Self {
+            id: gc.circ.id().clone(),
+            input_labels: gc
+                .data
+                .input_labels
+                .into_iter()
+                .map(InputLabels::from)
+                .collect::<Vec<InputLabels>>(),
+            encrypted_gates: gc
+                .data
+                .encrypted_gates
+                .into_iter()
+                .map(|gate| *gate.as_ref())
+                .flatten()
+                .collect::<Vec<Block>>(),
+            encoding: gc.data.encoding.and_then(|encoding| {
+                Some(
+                    encoding
+                        .into_iter()
+                        .map(OutputEncoding::from)
+                        .collect::<Vec<OutputEncoding>>(),
+                )
+            }),
+        }
+    }
+}
+
+impl crate::garble::GarbledCircuit<garble::Partial> {
+    pub fn from_msg(circ: Arc<Circuit>, msg: GarbledCircuit) -> Result<Self, crate::garble::Error> {
```

Comment:
Is this too computation intensive for a `TryFrom` implementation? Does it make sense to break it down into smaller pieces?

Reply:
Hmm now that I see it, I have changed my mind, lol. I agree, it is too clunky, and it does not make sense to create a wrapper type for both structs just to satisfy the `TryFrom` interface. Probably the best to leave it like it is.